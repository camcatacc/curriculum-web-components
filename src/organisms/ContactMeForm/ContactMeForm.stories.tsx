// Modules
import React, { useState } from "react";

// Storybook Modules
import { ComponentStory, ComponentMeta } from "@storybook/react";

// Elements
import ContactMeForm, { ContactForm, ContactFormButtonType } from "organisms/ContactMeForm/ContactMeForm";
import PageSize from "atoms/PageSize/PageSize";
import { Button } from "@material-ui/core";

// Default Props
const defaultForm = {
	name: "",
	surname: "",
	email: "",
	messageSubject: "",
	messageBody: ""
};

// Template
const Template: ComponentStory<typeof ContactMeForm> = (args) => {
	const [modalOpen, setModalOpen] = useState(false);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState("");
	const [form, setForm] = useState<ContactForm>({
		name: "",
		surname: "",
		email: "",
		messageSubject: "",
		messageBody: ""
	});

	function onFormChange(type: keyof ContactForm, newValue: string) {
		const copiedForm = { ...form };
		copiedForm[type] = newValue;
		setForm(copiedForm);
	}

	function onButtonClick(type: ContactFormButtonType) {
		setError("");
		switch (type) {
			case "cancel":
				setForm(defaultForm);
				handleModalClose();
				return;
			case "save":
				handleModalClose();
				return;
			case "send":
				handleSend();
				return;
			case "sendWithError":
				handleSend(true);
				return;
		}
	}

	function handleModalClose() {
		setModalOpen(false);
	}

	function handleModalOpen() {
		setModalOpen(true);
	}

	async function handleSend(newError?: boolean) {
		setLoading(true);
		await resolveAfterMilliseconds(1000);
		setLoading(false);
		if (newError) {
			setError("There has been an error");
			return;
		}
		setModalOpen(false);
	}

	function resolveAfterMilliseconds(milliseconds: number) {
		return new Promise((resolve) => setTimeout(() => resolve(true), milliseconds));
	}

	return (
		<PageSize>
			<Button onClick={handleModalOpen} variant="contained" color="secondary">
				Open modal
			</Button>
			<ContactMeForm
				openModal={modalOpen}
				onFormChange={onFormChange}
				onButtonClick={onButtonClick}
				loading={loading}
				handleModalClose={handleModalClose}
				error={error}
				form={form}
			/>
		</PageSize>
	);
};

// Variants
export const Standard = Template.bind({});
Standard.args = {};

// Export
export default {
	title: "Organisms/ContactMeForm",
	component: ContactMeForm
} as ComponentMeta<typeof ContactMeForm>;
