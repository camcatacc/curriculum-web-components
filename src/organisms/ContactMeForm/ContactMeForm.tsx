// Modules
import React from "react";

// Elements
import { Dialog, DialogTitle, DialogContent, DialogContentText, DialogActions, Button, TextField, Typography } from "@material-ui/core";
import ButtonWithState from "atoms/ButtonWithState/ButtonWithState";

// Styles
import * as Styled from "organisms/ContactMeForm/ContactMeForm.styled";

// Definitions
export interface ContactForm {
	name: string;
	surname: string;
	email: string;
	messageSubject: string;
	messageBody: string;
}

export type ContactFormButtonType = "cancel" | "send" | "save" | "sendWithError";
export interface ContactMeFormProps {
	openModal: boolean;
	handleModalClose: () => void;
	form: {
		name: string;
		surname: string;
		email: string;
		messageSubject: string;
		messageBody: string;
	};
	onFormChange: (type: keyof ContactForm, newValue: string) => void;
	onButtonClick: (type: ContactFormButtonType) => void;
	loading?: boolean;
	error?: string;
}

// Components
const ContactMeForm = ({ openModal, handleModalClose, form, onFormChange, onButtonClick, loading, error }: ContactMeFormProps) => (
	<Dialog open={openModal} onClose={handleModalClose} fullWidth>
		<DialogTitle>
			<Typography variant="h2" align="center">
				CONTACT ME
			</Typography>
		</DialogTitle>
		<DialogContent>
			<DialogContentText variant="body1" align="center">
				I will be glad to help ;)
			</DialogContentText>
			<div className="flex flex-col gap-3 mt-6">
				<TextField id="name" onChange={(ev) => onFormChange("name", ev.target.value)} value={form.name} variant="outlined" label="Name" />
				<TextField id="surname" onChange={(ev) => onFormChange("surname", ev.target.value)} value={form.surname} variant="outlined" label="Surname" />
				<TextField id="email" onChange={(ev) => onFormChange("email", ev.target.value)} value={form.email} variant="outlined" label="Email" />
				<TextField id="messageSubject" onChange={(ev) => onFormChange("messageSubject", ev.target.value)} value={form.messageSubject} variant="outlined" label="Message Subject" />
				<TextField
					id="messageBody"
					onChange={(ev) => onFormChange("messageBody", ev.target.value)}
					value={form.messageBody}
					variant="outlined"
					rows={5}
					multiline
					label="Message Body"
				/>
			</div>
		</DialogContent>
		<DialogActions>
			<div className="flex items-center justify-center flex-col md:flex-row w-full md:w-auto">
				<ButtonWithState variant="text" loading={loading} disabled={loading} onClick={() => onButtonClick("sendWithError")}>
					Send with error
				</ButtonWithState>
				<Button color="secondary" onClick={() => onButtonClick("cancel")}>
					Cancel
				</Button>
				<Button variant="outlined" color="secondary" onClick={() => onButtonClick("save")}>
					Finish it later
				</Button>
				<ButtonWithState variant="contained" loading={loading} disabled={loading} onClick={() => onButtonClick("send")}>
					Send
				</ButtonWithState>
			</div>
		</DialogActions>
		{error && (
			<Styled.Error variant="body2" align="right" color="error">
				{error}
			</Styled.Error>
		)}
	</Dialog>
);
export default ContactMeForm;
