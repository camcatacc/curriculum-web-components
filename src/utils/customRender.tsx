import { ThemeProvider } from "@material-ui/styles";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ReactElement } from "react";
import { getMaterialTheme } from "styles/materialUi";

export const customRender = (ui: ReactElement, renderOptions?: RenderOptions & { darkMode?: boolean }): RenderResult => {
	const Wrapper: React.FC = ({ children }) => {
		return <ThemeProvider theme={getMaterialTheme(renderOptions?.darkMode ? "dark" : "light")}>{children}</ThemeProvider>;
	};

	return render(ui, { wrapper: Wrapper, ...renderOptions });
};
