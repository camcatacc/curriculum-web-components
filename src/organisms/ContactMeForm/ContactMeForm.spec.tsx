// Modules
import React from "react";
import { fireEvent } from "@testing-library/react";
import { customRender } from "utils/customRender";

// Elements
import ContactMeForm from "organisms/ContactMeForm/ContactMeForm";

// Definitions
import type { ContactMeFormProps } from "organisms/ContactMeForm/ContactMeForm";

// Default Props
const defaultProps: ContactMeFormProps = {
	form: {
		email: "",
		messageBody: "",
		messageSubject: "",
		name: "",
		surname: ""
	},
	handleModalClose: jest.fn(),
	onButtonClick: jest.fn(),
	onFormChange: jest.fn(),
	openModal: true
};

// Tests
describe("ContactMeForm", () => {
	it("Renders as expected", () => {
		const { container } = customRender(<ContactMeForm {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});

	it("Calls onFormChange when name input changes", () => {
		const { getByLabelText } = customRender(<ContactMeForm {...defaultProps} />);
		const nameInput = getByLabelText("Name");
		fireEvent.change(nameInput, { target: { value: "David" } });
		expect(defaultProps.onFormChange).toHaveBeenCalled();
	});

	it("Calls onFormChange when surname input changes", () => {
		const { getByLabelText } = customRender(<ContactMeForm {...defaultProps} />);
		const surnameInput = getByLabelText("Surname");
		fireEvent.change(surnameInput, { target: { value: "Camacho" } });
		expect(defaultProps.onFormChange).toHaveBeenCalled();
	});

	it("Calls onFormChange when email input changes", () => {
		const { getByLabelText } = customRender(<ContactMeForm {...defaultProps} />);
		const emailInput = getByLabelText("Email");
		fireEvent.change(emailInput, { target: { value: "a@a.com" } });
		expect(defaultProps.onFormChange).toHaveBeenCalled();
	});

	it("Calls onFormChange when messageSubject input changes", () => {
		const { getByLabelText } = customRender(<ContactMeForm {...defaultProps} />);
		const subjectInput = getByLabelText("Message Subject");
		fireEvent.change(subjectInput, { target: { value: "Hello" } });
		expect(defaultProps.onFormChange).toHaveBeenCalled();
	});

	it("Calls onFormChange when messageBody input changes", () => {
		const { getByLabelText } = customRender(<ContactMeForm {...defaultProps} />);
		const messageInput = getByLabelText("Message Body");
		fireEvent.change(messageInput, { target: { value: "Hello, it's time to go" } });
		expect(defaultProps.onFormChange).toHaveBeenCalled();
	});
	it("Calls onButtonClick when Send with error is clicked", () => {
		const { getByText } = customRender(<ContactMeForm {...defaultProps} />);
		const button = getByText("Send with error");
		fireEvent.click(button);
		expect(defaultProps.onButtonClick).toHaveBeenCalled();
	});
	it("Calls onButtonClick when Cancel is clicked", () => {
		const { getByText } = customRender(<ContactMeForm {...defaultProps} />);
		const button = getByText("Cancel");
		fireEvent.click(button);
		expect(defaultProps.onButtonClick).toHaveBeenCalled();
	});
	it("Calls onButtonClick when Finish it later is clicked", () => {
		const { getByText } = customRender(<ContactMeForm {...defaultProps} />);
		const button = getByText("Finish it later");
		fireEvent.click(button);
		expect(defaultProps.onButtonClick).toHaveBeenCalled();
	});
	it("Calls onButtonClick when Send is clicked", () => {
		const { getByText } = customRender(<ContactMeForm {...defaultProps} />);
		const button = getByText("Send");
		fireEvent.click(button);
		expect(defaultProps.onButtonClick).toHaveBeenCalled();
	});
	it("Shows an error if passed", () => {
		const errorProp = "There was an error";
		const { getByText } = customRender(<ContactMeForm {...defaultProps} error={errorProp} />);
		const error = getByText(errorProp);
		expect(error).toBeInTheDocument();
	});
});
