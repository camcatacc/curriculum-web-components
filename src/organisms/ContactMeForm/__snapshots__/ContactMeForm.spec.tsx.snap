// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMeForm Renders as expected 1`] = `
<div
  aria-hidden="true"
/>
`;
